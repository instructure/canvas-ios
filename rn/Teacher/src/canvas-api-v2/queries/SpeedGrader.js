//
// This file is part of Canvas.
// Copyright (C) 2019-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

import gql from 'graphql-tag'

export default gql`query SpeedGrader($assignmentID: ID!, $states: [SubmissionState!], $late: Boolean, $scoredMoreThan: Float, $scoredLessThan: Float, $sectionIDs: [ID!], $gradingStatus: SubmissionGradingStatus) {
  assignment(id: $assignmentID) {
    gradeGroupStudentsIndividually
    anonymizeStudents
    submissionTypes

    groupSet {
      id
    }

    course {
      groups: groupsConnection {
        edges {
          group: node {
            id: _id
            name
            members: membersConnection {
              edges {
                member: node {
                  user {
                    id: _id
                  }
                }
              }
            }
          }
        }
      }
    }

    submissions: submissionsConnection(
      filter: {
        states: $states
        late: $late
        scoredMoreThan: $scoredMoreThan
        scoredLessThan: $scoredLessThan
        sectionIds: $sectionIDs
        gradingStatus: $gradingStatus
      }
      orderBy: {
        field: username
      }
    ) {
      edges {
        submission: node {
          id
          _id
          missing
          late
          state
          submittedAt
          attempt

          attachments {
            id
            _id
            url
            displayName
            thumbnailUrl
            mimeClass
          }

          user {
            id: _id
            name
            avatarUrl
          }

          comments: commentsConnection(filter: { allComments: true }) {
            nodes {
              _id
              createdAt
              comment

              author {
                _id
                name
                avatarUrl
              }

              mediaObject {
                _id
                mediaType
                title

                mediaSources {
                  url
                }
              }
            }
          }

          submissionHistory: submissionHistoriesConnection {
            edges {
              submission: node {
                rootId
                missing
                late
                state
                submittedAt
                attempt

                attachments {
                  id
                  _id
                  url
                  displayName
                  thumbnailUrl
                  mimeClass
                }
              }
            }
          }
        }
      }
    }
  }
}`
