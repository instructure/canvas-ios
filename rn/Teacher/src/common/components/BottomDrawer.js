//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import React, { Component } from 'react'
import {
  Dimensions,
  Animated,
  View,
  requireNativeComponent,
} from 'react-native'
import Interactable from 'react-native-interactable'
import Button from 'react-native-button'
import i18n from 'format-message'
import DrawerState, {
  HANDLE_PADDING_BOTTOM,
  HANDLE_INVISIBLE_TOP_PADDING,
  HANDLE_HEIGHT,
  HANDLE_BAR_HEIGHT,
  HANDLE_PADDING_TOP,
  HANDLE_VISIBLE_PADDING,
} from '../../modules/speedgrader/utils/drawer-state'
import { createStyleSheet } from '../stylesheet'

let { height, width } = Dimensions.get('window')

const DrawerHandle = requireNativeComponent('DrawerHandleView')

export type Snap = 0 | 1 | 2
const SNAP_DEFAULT_INDEX = 0

type Props = {
  containerWidth?: number,
  containerHeight?: number,
  children?: React$Node,
  renderHandleContent?: () => any,
  drawerState: DrawerState,
}

type State = {
  height: number,
  width: number,
}

export default class BottomDrawer extends Component<Props, State> {
  state: State = {
    height: this.props.containerHeight || height,
    width: this.props.containerWidth || width,
  }
  _deltaY: Animated.Value = this.props.drawerState.deltaY
  drawer: InteractableView

  shouldComponentUpdate (nextProps: Props, nextState: State) {
    if (nextProps.containerHeight === 0) return false
    return true
  }

  UNSAFE_componentWillReceiveProps (nextProps: Props) {
    if (nextProps.containerHeight !== this.props.containerHeight || nextProps.containerWidth !== this.props.containerWidth) {
      this.setState({ height: nextProps.containerHeight, width: nextProps.containerWidth }, () => {
        if (this.drawer) {
          this.drawer.snapTo({ index: this.props.drawerState.currentSnap })
        }
      })
    }
  }

  UNSAFE_componentWillMount () {
    this.props.drawerState.registerDrawer(this)
  }

  componentWillUnmount () {
    this.props.drawerState.unregisterDrawer(this)
  }

  open = () => {
    if (this.props.drawerState.currentSnap === SNAP_DEFAULT_INDEX) {
      this.drawer.snapTo({ index: 1 })
    }
  }

  cycle = () => {
    let index = (this.props.drawerState.currentSnap + 1) % 3
    this.drawer.snapTo({ index })
  }

  onDrag = ({ nativeEvent }: { nativeEvent: Object }) => {
    if (nativeEvent.state === 'start') {
      this.props.drawerState.dragBegan()
    }
  }

  onSnap = (e: any) => {
    this.props.drawerState.drawerDidSnap(this, e.nativeEvent.index)
  }

  snapTo = (snap: Snap, animated: boolean = true) => {
    if (this.drawer) {
      this.drawer.snapTo({ index: snap, animated: animated })
    }
  }

  captureDrawer = (d: any) => {
    this.drawer = d
  }

  getSnapPoints = () => {
    const drawerState = this.props.drawerState
    return [0, 1, 2]
      .map(snap => ({
        y: drawerState.drawerSnapPoint(snap, this.state.height),
      }))
  }

  render () {
    const cycleText = i18n('Open Drawer')
    const snapPointStates = [
      i18n('Full screen'),
      i18n('Half screen'),
      i18n('Closed'),
    ]

    let width = Math.min(700, this.state.width)
    let position = (this.state.width - width) / 2

    const snap = this.getSnapPoints().map(point => point.y) // [0, 300, 600]
    const maxSnap = snap[snap.length - 1]
    const heights = [ snap[1], snap[1], maxSnap ] // [300, 300, 600]
    const clampedHeight = this._deltaY.interpolate({
      inputRange: snap,
      outputRange: heights.map(h => h + HANDLE_PADDING_BOTTOM),
    })
    const bottoms = [ -snap[1], 0, 0 ] // [-300, 0, 0]
    const clampedBottom = this._deltaY.interpolate({
      inputRange: snap,
      outputRange: bottoms,
    })
    const hideContentForA11y = this._deltaY.interpolate({
      inputRange: [0, 1, maxSnap],
      outputRange: [0, 1, 1],
    })

    return (
      <View style={[styles.drawer, { left: position, right: position }]}>
        <Interactable.View
          ref={this.captureDrawer}
          onSnap={this.onSnap}
          verticalOnly={true}
          snapPoints={this.getSnapPoints()}
          initialPosition={this.getSnapPoints()[this.props.drawerState.currentSnap]}
          animatedValueY={this._deltaY}
          animatedValueX={new Animated.Value(0)}
          style={styles.handle}
          onDrag={this.onDrag}
        />
        <Animated.View
          style={[styles.drawerContent, {
            height: clampedHeight,
            top: clampedBottom,
          }]}
        >
          <View style={styles.drawerBackground} />
          <DrawerHandle
            style={[{ marginTop: -HANDLE_INVISIBLE_TOP_PADDING }]}
          >
            <Button
              accessible
              accessibilityTraits={['header', 'button']}
              onPress={this.cycle} testID='bottom-drawer.cycle'
            >
              <View style={styles.handleBar} accessibilityLabel={`${cycleText} ${snapPointStates[this.props.drawerState.currentSnap]}`}/>
            </Button>
            <View style={styles.handleContent}>{
              this.props.renderHandleContent
                ? this.props.renderHandleContent()
                : undefined
            }</View>
          </DrawerHandle>
          <Animated.View style={{ flex: 1, opacity: hideContentForA11y }}>
            {this.props.children}
          </Animated.View>
        </Animated.View>
      </View>
    )
  }
}

const styles = createStyleSheet((colors, vars) => ({
  drawer: {
    transform: [{ rotate: '180deg' }],
    position: 'absolute',
    bottom: 0,
    height: HANDLE_PADDING_BOTTOM,
  },
  drawerContent: {
    transform: [{ rotate: '180deg' }],
    position: 'absolute',
    left: 0,
    right: 0,
  },
  drawerBackground: {
    backgroundColor: colors.backgroundLightest,
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    shadowColor: 'black',
    shadowOffset: { width: 0, height: 0 },
    shadowRadius: 3,
    shadowOpacity: 0.3,
    borderTopLeftRadius: 12,
    borderTopRightRadius: 12,
  },
  handle: {
    position: 'absolute',
    left: vars.padding,
    right: vars.padding,
    height: HANDLE_HEIGHT,
  },
  handleBar: {
    width: 40,
    height: HANDLE_BAR_HEIGHT,
    backgroundColor: colors.backgroundDark,
    borderRadius: 5,
    marginTop: HANDLE_PADDING_TOP,
    alignSelf: 'center',
  },
  handleContent: {
    marginTop: HANDLE_VISIBLE_PADDING,
  },
}))
