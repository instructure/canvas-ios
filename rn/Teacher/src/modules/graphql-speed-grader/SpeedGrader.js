//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/* eslint-disable flowtype/require-valid-file-annotation */

import React, { Component } from 'react'
import {
  View,
  StyleSheet,
  ActivityIndicator,
  FlatList,
  Dimensions,
  NativeModules,
  Button,
} from 'react-native'
import SubmissionGrader from './SubmissionGrader'
import type {
  AsyncSubmissionsDataProps,
  SubmissionDataProps,
} from '../submissions/list/submission-prop-types'
import { updateBadgeCounts } from '../tabbar/badge-counts'
import Screen from '../../routing/Screen'
import Navigator from '../../routing/Navigator'
import DrawerState, { type DrawerPosition } from '../speedgrader/utils/drawer-state'
import Tutorial from '../speedgrader/components/Tutorial'
import i18n from 'format-message'
import Images from '../../images'
// import shuffle from 'knuth-shuffle-seeded'
import { Title } from '../../common/text'
import CommentInput from '../speedgrader/comments/CommentInput'
import A11yGroup from '../../common/components/A11yGroup'
import { graphql } from 'react-apollo'
import query from '../../canvas-api-v2/queries/SpeedGrader'

const { NativeAccessibility } = NativeModules

type State = {
  size: { width: number, height: number },
  currentStudentID: ?string,
  drawerInset: number,
  hasScrolledToInitialSubmission: boolean,
  hasSetInitialDrawerPosition: boolean,
  submissions?: Array<SubmissionDataProps>,
}

const PAGE_GUTTER_HALF_WIDTH = 10.0

export class SpeedGrader extends Component<SpeedGraderProps, State> {
  props: SpeedGraderProps
  state: State
  _flatList: ?FlatList
  scrollView: ?{ setNativeProps: (Object) => void }
  hasRenderedBody = false

  static drawerState = new DrawerState()
  static defaultProps = {
    onDismiss: () => {},
    drawerPosition: 0,
  }

  constructor (props: SpeedGraderProps) {
    super(props)

    const { height, width } = Dimensions.get('window')
    const position = SpeedGrader.drawerState.currentSnap
    this.state = {
      size: {
        width: width + PAGE_GUTTER_HALF_WIDTH + PAGE_GUTTER_HALF_WIDTH,
        height,
      },
      currentStudentID: props.userID,
      currentPageIndex: props.studentIndex,
      drawerInset: SpeedGrader.drawerState.drawerHeight(position, height),
      hasScrolledToInitialSubmission: false,
      hasSetInitialDrawerPosition: false,
      scrollEnabled: true,
    }
    SpeedGrader.drawerState.registerDrawer(this)
  }

  // DrawerObserver
  snapTo = (position: DrawerPosition) => {
    this.setState({ drawerInset: SpeedGrader.drawerState.drawerHeight(position, this.state.size.height) })
  }

  componentWillUnmount () {
    SpeedGrader.drawerState.unregisterDrawer(this)
    SpeedGrader.drawerState.snapTo(0, false)
    updateBadgeCounts()
  }

  onLayout = (event: any) => {
    let viewableAreaChanged = false
    const { width, height } = event.nativeEvent.layout
    if (height !== 0 && (width !== this.state.size.width || height !== this.state.size.height)) {
      viewableAreaChanged = true
      this.setState({ size: { width, height } })
    }

    this.setState((prevState, props) => {
      let nextState = prevState

      if (this._flatList && !prevState.hasSetInitialDrawerPosition) {
        if (this.getInitialTabIndex() >= 0) {
          SpeedGrader.drawerState.snapTo(this.getInitialTabIndex(), false)
        }
        nextState = { ...nextState, hasSetInitialDrawerPosition: true }
      }

      const scrollToCurrentPageIndex = !prevState.hasScrolledToInitialSubmission || viewableAreaChanged

      if (this._flatList && scrollToCurrentPageIndex) {
        this._flatList.scrollToOffset({ animated: false, offset: nextState.size.width * this.state.currentPageIndex })
        nextState = { ...nextState, hasScrolledToInitialSubmission: true }
      }

      return nextState
    })
  }

  _captureFlatList = (list: FlatList) => {
    this._flatList = list
  }

  renderItem = ({ item, index }: { item: SubmissionItem, index: number }) => {
    const isCurrentStudent = this.state.currentStudentID
      ? this.state.currentStudentID === item.user.id
      : index === 0

    let group = this.props.isGroupGradedAssignment
      ? this.props.groups.find(({ members }) => {
        return members.edges.find(({ member }) => member.user.id === item.user.id)
      })
      : null

    return <A11yGroup style={[styles.page, this.state.size]}>
      <SubmissionGrader
        isCurrentStudent={isCurrentStudent}
        drawerState={SpeedGrader.drawerState}
        courseID={this.props.courseID}
        assignmentID={this.props.assignmentID}
        userID={item.user.id}
        submissionID={item.id}
        closeModal={this.dismiss}
        submission={item}
        assignmentSubmissionTypes={this.props.assignmentSubmissionTypes}
        isModeratedGrading={this.props.isModeratedGrading}
        navigator={this.props.navigator}
        drawerInset={this.state.drawerInset}
        gradeSubmissionWithRubric={this.props.gradeSubmissionWithRubric}
        selectedTabIndex={this.getInitialTabIndex()}
        setScrollEnabled={(value) => {
          this.scrollView.setNativeProps({ scrollEnabled: value })
        }}
        group={group}
        anonymousGrading={this.props.anonymousGrading}
        assignment={this.props.assignment}
      />
    </A11yGroup>
  }

  dismiss = () => {
    this.props.navigator.dismiss()
    this.props.onDismiss()
  }

  scrollEnded = (event: Object) => {
    const index = event.nativeEvent.contentOffset.x / this.state.size.width
    if (index !== this.state.currentPageIndex) {
      CommentInput.persistentComment.text = ''
    }
    this.setState({ currentPageIndex: index })
    const submission = (this.props.submissions || [])[index]
    if (submission) {
      const currentStudentID = submission.userID
      if (currentStudentID !== this.state.currentStudentID) {
        this.setState({ currentStudentID })
      }
    }
  }

  getItemLayout = (data: ?any, index: number) => ({
    length: this.state.size.width,
    offset: this.state.size.width * index,
    index,
  })

  getInitialTabIndex = () => {
    if (this.props.pushNotification && this.props.pushNotification.alert.toLowerCase().startsWith(i18n('submission comment'))) {
      return 1
    }
    return -1
  }

  renderBody = () => {
    if (this.props.pending) {
      return (
        <View
          accessible
          accessibilityLabel={i18n('In Progress')}
          style={styles.loadingWrapper}
        >
          <ActivityIndicator />
        </View>
      )
    }

    // This is rare, but if it does happen, SpeedGrader gets into a loading loop and the app basically explodes
    // jk, it loads forever and you can't do anything else unless you force close the app
    if (this.props.submissions.length === 0) {
      return (
        <View
          accessible
          accessibilityLabel={i18n('No Submissions to Display')}
          style={styles.loadingWrapper}
        >
          <Title style={{ margin: 16, textAlign: 'center' }}>{i18n("It seems there aren't any valid submissions to grade.")}</Title>
          <Button onPress={this.dismiss} title={i18n('Close')} />
        </View>
      )
    }

    if (!this.hasRenderedBody) {
      this.hasRenderedBody = true
      NativeAccessibility.refresh()
    }

    return (
      <FlatList
        ref={this._captureFlatList}
        keyboardShouldPersistTaps='handled'
        onLayout={this.onLayout}
        data={this.props.submissions}
        renderItem={this.renderItem}
        windowSize={5}
        initialNumToRender={null}
        horizontal
        pagingEnabled
        getItemLayout={this.getItemLayout}
        showsHorizontalScrollIndicator={false}
        onMomentumScrollEnd={this.scrollEnded}
        style={{ marginLeft: -PAGE_GUTTER_HALF_WIDTH, marginRight: -PAGE_GUTTER_HALF_WIDTH }}
        ref={(e) => { this.scrollView = e }}
      />
    )
  }

  render () {
    let tutorials = [{
      id: 'swipe-tutorial',
      text: i18n('Swipe left or right to view other student submissions'),
      image: Images.speedGrader.swipe,
    }]

    if (this.props.hasRubric) {
      tutorials.push({
        id: 'long-press-tutorial',
        text: i18n('Tap and hold a rubric number to see its description'),
        image: Images.speedGrader.longPress,
      })
    }

    return (
      <Screen
        navBarHidden
        noRotationInVerticallyCompact
      >
        <View style={styles.speedGrader}>
          { this.renderBody() }
          <Tutorial
            tutorials={tutorials}
          />
        </View>
      </Screen>
    )
  }
}

const styles = StyleSheet.create({
  page: {
    paddingLeft: 10,
    paddingRight: 10,
    overflow: 'hidden',
  },
  speedGrader: {
    flex: 1,
  },
  loadingWrapper: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
})

function props (props) {
  if (props.data.loading) {
    return {
      pending: true,
      hasRubric: false,
    }
  }

  let assignment = props.data.assignment
  let rubric = assignment.rubric
  let submissions = assignment.submissions.edges.map(({ submission }) => submission)
  let course = assignment.course
  let groups = course.groups.edges.map(({ group }) => group)
  let quiz = assignment.quiz
  return {
    pending: false,
    hasRubric: rubric != null,
    assignment,
    submissions,
    isGroupGradedAssignment: (
      assignment.groupSet?.id != null &&
      !assignment.gradeGroupStudentsIndividually
    ),
    groups,
    anonymousGrading: quiz?.anonymousSubmissions ?? assignment.anonymizeStudents ?? false,
  }
}

export default graphql(query, {
  options: ({ assignmentID, filter }) => {
    return {
      variables: {
        assignmentID,
        ...filter,
      },
    }
  },
  fetchPolicy: 'cache-and-network',
  props,
})(SpeedGrader)

type SubmissionItem = {
  key: string,
  submission: SubmissionDataProps,
}
type RoutingProps = {
  courseID: string,
  assignmentID: string,
  userID: string,
  filter?: Function,
  studentIndex: number,
}
type SpeedGraderActionProps = {
  refreshSubmissions: Function,
  refreshSubmissionSummary: Function,
  refreshEnrollments: Function,
  refreshAssignment: Function,
  refreshGroupsForCourse: Function,
  gradeSubmissionWithRubric: Function,
}
type SpeedGraderDataProps = {
  submissionEntities: SubmissionsState,
  groupAssignment: ?{ groupCategoryID: string, gradeIndividually: boolean },
  assignmentSubmissionTypes: Array<SubmissionType>,
  isModeratedGrading: boolean,
  hasRubric: boolean,
  hasAssignment: boolean,
} & AsyncSubmissionsDataProps

type SpeedGraderProps
  = RoutingProps
  & SpeedGraderActionProps
  & SpeedGraderDataProps
  & RefreshProps
  & PushNotificationProps
  & { navigator: Navigator, onDismiss: Function }
