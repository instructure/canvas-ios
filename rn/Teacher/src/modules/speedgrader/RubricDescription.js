//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import React, { Component } from 'react'
import { View, StyleSheet, Text } from 'react-native'
import { connect } from 'react-redux'
import i18n from 'format-message'
import Navigator from '../../routing/Navigator'
import Screen from '../../routing/Screen'
import CanvasWebView from '../../common/components/CanvasWebView'

export class RubricDescription extends Component<*> {
  renderLongDescription () {
    if (!this.props.description || this.props.description.length === 0) {
      return (
        <View style={styles.emptyState}>
          <Text style={styles.emptyStateText}>{i18n('There currently is no long description for this item.')}</Text>
        </View>
      )
    }
    return (
      <View style={styles.container}>
        <CanvasWebView html={this.props.description} automaticallySetHeight navigator={this.props.navigator}/>
      </View>
    )
  }

  render () {
    const description = this.renderLongDescription()
    return (
      <Screen
        title={i18n('Long Description')}
      >
        { description }
      </Screen>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    paddingTop: 24,
    paddingHorizontal: 16,
  },
  emptyState: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    paddingHorizontal: 32,
  },
  emptyStateText: {
    color: '#2D3B45',
    fontSize: 16,
    lineHeight: 19,
    textAlign: 'center',
  },
})

export function mapStateToProps (state: AppState, ownProps: RubricDescriptionOwnProps): RubricDescriptionDataProps {
  let rubric = state.entities.assignments[ownProps.assignmentID].data.rubric
  if (!rubric) {
    return { description: '' }
  }

  return {
    description: rubric.find(r => r.id === ownProps.rubricID).long_description,
  }
}

const Connected = connect(mapStateToProps)(RubricDescription)
export default (Connected: any)

type RubricDescriptionOwnProps = {
  assignmentID: string,
  rubricID: string,
  navigator: Navigator,
}

type RubricDescriptionDataProps = {
  description: string,
}
