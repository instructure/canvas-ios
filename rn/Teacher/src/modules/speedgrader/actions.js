//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import { createAction } from 'redux-actions'
import canvas from '../../canvas-api'

export const SpeedGraderActions = (api: CanvasApi): * => ({
  excuseAssignment: createAction('submission.excuse', (courseID: string, assignmentID: string, userID: string, submissionID: ?string) => ({
    promise: api.gradeSubmission(courseID, assignmentID, userID, {
      excuse: true,
    }),
    submissionID,
    assignmentID,
  })),
  selectSubmissionFromHistory: createAction('submission.selectFromHistory', (submissionID: string, index: number) => ({
    submissionID,
    index,
  })),
  selectFile: createAction('submission.selectFile', (submissionID: string, index: number) => ({
    submissionID,
    index,
  })),
  gradeSubmission: createAction('submission.grade', (courseID: string, assignmentID: string, userID: string, submissionID: ?string, grade: string) => ({
    promise: api.gradeSubmission(courseID, assignmentID, userID, {
      posted_grade: grade,
    }),
    submissionID,
    assignmentID,
    handlesError: true,
  })),
  gradeSubmissionWithRubric: createAction('submission.gradeWithRubric', (courseID: string, assignmentID: string, userID: string, submissionID: ?string, rubricParams: { [string]: RubricAssessment }) => ({
    promise: api.gradeSubmissionWithRubric(courseID, assignmentID, userID, rubricParams),
    submissionID,
    assignmentID,
    rubricAssessment: rubricParams,
  })),
})

export default (SpeedGraderActions(canvas): *)
