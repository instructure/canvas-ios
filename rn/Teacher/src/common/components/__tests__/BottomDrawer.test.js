//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow
import React from 'react'
import { Text } from 'react-native'
import BottomDrawer from '../BottomDrawer'
import renderer from 'react-test-renderer'
import setProps from '../../../../test/helpers/setProps'
import explore from '../../../../test/helpers/explore'
import { SpeedGrader } from '../../../modules/speedgrader/SpeedGrader'

jest
  .mock('react-native-interactable', () => ({
    View: 'Interactable.View',
  }))
  .mock('react-native-button', () => 'Button')

describe('BottomDrawer', () => {
  beforeEach(() => jest.clearAllMocks())
  const state = SpeedGrader.drawerState

  it('renders any children', () => {
    let tree = renderer.create(
      <BottomDrawer currentSnap={2} drawerState={state}>
        <Text>Yo yo yo</Text>
      </BottomDrawer>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('sets the new height and width state when props change', () => {
    let tree = renderer.create(
      <BottomDrawer containerHeight={200} containerWidth={100} currentSnap={2} drawerState={state}>
        <Text>yo yo yo</Text>
      </BottomDrawer>
    )
    let instance = tree.getInstance()
    expect(instance.state.height).toEqual(200)
    expect(instance.state.width).toEqual(100)

    setProps(tree, { containerHeight: 240, containerWidth: 200 })
    instance = tree.getInstance()
    expect(instance.state.height).toEqual(240)
    expect(instance.state.width).toEqual(200)
  })

  it('calls snapTo when currentSnap is 2', () => {
    let tree = renderer.create(
      <BottomDrawer currentSnap={2} drawerState={state}>
        <Text>Yo yo yo</Text>
      </BottomDrawer>
    )
    let instance = tree.getInstance()
    let snapTo = jest.fn()
    instance.drawer = { snapTo }
    instance.open()
    expect(snapTo).toHaveBeenCalledWith({ index: 1 })
  })

  it('doesnt call snapTo when the currentSnap is not 0', () => {
    state.currentSnap = 2
    let tree = renderer.create(
      <BottomDrawer drawerState={state}>
        <Text>yo yo yo</Text>
      </BottomDrawer>
    )
    let instance = tree.getInstance()
    instance.drawer = { snapTo: jest.fn() }
    instance.open()
    expect(instance.drawer.snapTo).not.toHaveBeenCalled()
  })

  it('returns 3 snap points from getSnapPoints', () => {
    let tree = renderer.create(
      <BottomDrawer drawerState={state}>
        <Text>yo yo yo</Text>
      </BottomDrawer>
    )
    let instance = tree.getInstance()
    let snapPoints = instance.getSnapPoints()
    expect(snapPoints[0].y).toBeLessThan(snapPoints[1].y)
    expect(snapPoints[1].y).toBeLessThan(snapPoints[2].y)
    expect(snapPoints.length).toEqual(3)
  })

  // because we setState after a cycle, and the drawer is a ref
  // the drawer attribute on the instance gets wiped out after every
  // call to cycle so we are just going to do it with 3 different trees
  it('cycles through snaps when the handle button is pressed', () => {
    state.currentSnap = 0
    let tree = renderer.create(
      <BottomDrawer drawerState={state}>
        <Text>yo yo yo</Text>
      </BottomDrawer>
    )
    let instance = tree.getInstance()
    let button = explore(tree.toJSON()).selectByID('bottom-drawer.cycle') || {}
    let snapTo = jest.fn()
    instance.drawer = { snapTo }
    button.props.onPress()
    expect(snapTo).toHaveBeenLastCalledWith({ index: 1 })

    state.currentSnap = 1
    tree = renderer.create(
      <BottomDrawer drawerState={state}>
        <Text>yo yo yo</Text>
      </BottomDrawer>
    )
    instance = tree.getInstance()
    button = explore(tree.toJSON()).selectByID('bottom-drawer.cycle') || {}
    snapTo = jest.fn()
    instance.drawer = { snapTo }
    button.props.onPress()
    expect(snapTo).toHaveBeenLastCalledWith({ index: 2 })

    state.currentSnap = 2
    tree = renderer.create(
      <BottomDrawer currentSnap={0} drawerState={state}>
        <Text>yo yo yo</Text>
      </BottomDrawer>
    )
    instance = tree.getInstance()
    button = explore(tree.toJSON()).selectByID('bottom-drawer.cycle') || {}
    snapTo = jest.fn()
    instance.drawer = { snapTo }
    button.props.onPress()
    expect(snapTo).toHaveBeenLastCalledWith({ index: 0 })
  })
})
