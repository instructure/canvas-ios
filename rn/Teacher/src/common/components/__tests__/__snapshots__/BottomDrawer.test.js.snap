// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomDrawer renders any children 1`] = `
<View
  style={
    Array [
      Object {
        "bottom": 0,
        "height": 52,
        "position": "absolute",
        "transform": Array [
          Object {
            "rotate": "180deg",
          },
        ],
      },
      Object {
        "left": 25,
        "right": 25,
      },
    ]
  }
>
  <Interactable.View
    animatedValueY={0}
    initialPosition={
      Object {
        "y": 0,
      }
    }
    onDrag={[Function]}
    onSnap={[Function]}
    snapPoints={
      Array [
        Object {
          "y": 0,
        },
        Object {
          "y": 615,
        },
        Object {
          "y": 1212,
        },
      ]
    }
    style={
      Object {
        "height": 78,
        "left": 16,
        "position": "absolute",
        "right": 16,
      }
    }
    verticalOnly={true}
  />
  <View
    style={
      Object {
        "height": 667,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": -615,
        "transform": Array [
          Object {
            "rotate": "180deg",
          },
        ],
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "white",
          "borderTopLeftRadius": 12,
          "borderTopRightRadius": 12,
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 3,
          "top": 0,
        }
      }
    />
    <DrawerHandleView
      style={
        Array [
          Object {
            "marginTop": -16,
          },
        ]
      }
    >
      <Button
        accessibilityTraits={
          Array [
            "header",
            "button",
          ]
        }
        accessible={true}
        onPress={[Function]}
        testID="bottom-drawer.cycle"
      >
        <View
          accessibilityLabel="Open Drawer Full screen"
          style={
            Object {
              "alignSelf": "center",
              "backgroundColor": "darkgray",
              "borderRadius": 5,
              "height": 4,
              "marginTop": 22,
              "width": 40,
            }
          }
        />
      </Button>
      <View
        style={
          Object {
            "marginTop": 6,
          }
        }
      />
    </DrawerHandleView>
    <View
      style={
        Object {
          "flex": 1,
          "opacity": 0,
        }
      }
    >
      <Text>
        Yo yo yo
      </Text>
    </View>
  </View>
</View>
`;
